File: Frontend/src/lib/currency.ts
/**
 * Single source of truth for money formatting.
 *
 * Two different kinds of money exist in this product and they do NOT share a
 * setting:
 *  - What Faro costs is priced in USD for everyone. That lives in Stripe and is
 *    rendered by the billing panel, never through here.
 *  - What the CUSTOMER's money is worth — inventory value, unit costs, a purchase
 *    order's total — is in whatever currency that business trades in. That is the
 *    tenant setting this module reads.
 *
 * The active currency is a module-level value rather than a hook return, and
 * `CurrencyProvider` sets it once on boot. That is deliberate: `formatMoney` is
 * called from ~200 places, many of them inside chart option builders and export
 * helpers that are not React components and cannot call a hook. Threading a
 * context through all of them would have meant touching every call site to change
 * one symbol.
 *
 * Two rules kept from a real bug:
 *  - `formatMoney` never abbreviates. Scaling to millions rendered ₡25,430 as
 *    "₡0.0M" — the product reporting near-zero for money the user actually had
 *    tied up in stock. SMB amounts are typically five figures.
 *  - `formatMoneyCompact` exists only for KPI tiles where space is tight, and it
 *    falls back to the full amount below 10,000 rather than showing 0.
 */

export interface CurrencyInfo {
  code:     string
  symbol:   string
  locale:   string
  decimals: number
}

/** Costa Rica is the anchor market, so a tenant that never chose keeps reading
 *  colones exactly as it did before this setting existed. Mirrors DEFAULT_CODE
 *  in backend/api/v1/currency.py. */
const DEFAULT: CurrencyInfo = {
  code: 'CRC', symbol: '₡', locale: 'es-CR', decimals: 0,
}

let active: CurrencyInfo = DEFAULT

/** Called by CurrencyProvider once the tenant's setting has loaded. */
export function setActiveCurrency(info: Partial<CurrencyInfo> | null | undefined) {
  if (!info?.code || !info.symbol) return
  active = {
    code: info.code,
    symbol: info.symbol,
    locale: info.locale || DEFAULT.locale,
    decimals: typeof info.decimals === 'number' ? info.decimals : 2,
  }
}

export function activeCurrency(): CurrencyInfo {
  return active
}

// Kept as live getters, not frozen constants: the old module exported
// `CURRENCY_CODE` / `CURRENCY_SYMBOL` as literals, and anything still importing
// them must follow the tenant's choice rather than a value captured at import.
export const getCurrencyCode   = () => active.code
export const getCurrencySymbol = () => active.symbol

/**
 * Full amount, no abbreviation. Use this everywhere by default.
 *
 * Honours the currency's declared `decimals`: it used to round unconditionally,
 * so a tenant on USD or MXN — both of which declare 2 — read "$8" for a unit cost
 * of $8.49. A 0-decimal currency (CRC, COP, CLP, ARS) still goes through the
 * exact same `Math.round(...).toLocaleString(locale)` as before, so the anchor
 * market renders byte-for-byte what it did.
 */
export function formatMoney(n: number | null | undefined): string {
  if (n == null || Number.isNaN(n)) return '—'
  const d = active.decimals
  if (!(d > 0)) return `${active.symbol}${Math.round(n).toLocaleString(active.locale)}`
  return `${active.symbol}${n.toLocaleString(active.locale, {
    minimumFractionDigits: d, maximumFractionDigits: d,
  })}`
}

/**
 * Abbreviated form for KPI tiles only. Never abbreviates below 10,000, so a real
 * five-figure amount can never collapse to "₡0.0M".
 */
export function formatMoneyCompact(n: number | null | undefined): string {
  if (n == null || Number.isNaN(n)) return '—'
  const abs = Math.abs(n)
  if (abs >= 1_000_000) return `${active.symbol}${(n / 1_000_000).toFixed(1)}M`
  if (abs >= 10_000) return `${active.symbol}${Math.round(n / 1_000)}K`
  return formatMoney(n)
}

/**
 * What Faro itself costs. Always USD, never the tenant's currency: the plans are
 * the same price for every customer, and showing "₡649" for a 649-dollar plan
 * would understate it by roughly five hundred times.
 */
export function formatPlanPrice(usd: number): string {
  return `$${usd.toLocaleString('en-US')}`
}

File: backend/api/v1/currency.py
"""Which currency this company's money is shown in.

Two different kinds of money live in this product and they must not share a
setting:

  · What Faro COSTS is priced in USD — the plans are the same price for everyone,
    and that lives in Stripe (backend/billing), not here.
  · What the customer's OWN money is worth — inventory value, unit costs, a
    purchase order's total — is in whatever currency that business actually
    trades in, and only they know which.

So this is a tenant-level setting, not a per-user one: it describes the
company's books, not a display preference. Two people in the same company
looking at the same purchase order must see the same number with the same symbol.

Changing it relabels existing figures; it does NOT convert them. Faro stores what
it was given and never guesses an exchange rate, so switching the currency on a
tenant that already has costs loaded is a mislabel, not a conversion — which is
why the endpoint is admin-only and the UI says so.
"""
import logging

from fastapi import APIRouter, Depends
from pydantic import BaseModel

from backend.auth.guards import CurrentUser, get_current_user, require_role
from backend.errors import AppError
from backend.schemas.common import ok
from backend.tenants.service import get_settings, update_settings

router = APIRouter(prefix="/tenant/currency", tags=["currency"])
log = logging.getLogger(__name__)

# ISO 4217 codes this deployment supports. Kept deliberately short: every entry
# needs a symbol and a locale the frontend can format with, and an unsupported
# code would render as a bare number with no symbol at all.
SUPPORTED = {
    "CRC": {"symbol": "₡", "locale": "es-CR", "decimals": 0},
    "USD": {"symbol": "$",      "locale": "en-US", "decimals": 2},
    "MXN": {"symbol": "$",      "locale": "es-MX", "decimals": 2},
    "GTQ": {"symbol": "Q",      "locale": "es-GT", "decimals": 2},
    "COP": {"symbol": "$",      "locale": "es-CO", "decimals": 0},
    "PAB": {"symbol": "B/.",    "locale": "es-PA", "decimals": 2},
    "HNL": {"symbol": "L",      "locale": "es-HN", "decimals": 2},
    "NIO": {"symbol": "C$",     "locale": "es-NI", "decimals": 2},
    "DOP": {"symbol": "RD$",    "locale": "es-DO", "decimals": 2},
    "PEN": {"symbol": "S/",     "locale": "es-PE", "decimals": 2},
    "CLP": {"symbol": "$",      "locale": "es-CL", "decimals": 0},
    "ARS": {"symbol": "$",      "locale": "es-AR", "decimals": 0},
    "EUR": {"symbol": "€", "locale": "es-ES", "decimals": 2},
}

# Costa Rica is the anchor market, so an existing tenant that never chose keeps
# reading colones exactly as it did before this setting existed.
DEFAULT_CODE = "CRC"


class CurrencyUpdate(BaseModel):
    code: str


def currency_of(tenant_id: str) -> dict:
    code = (get_settings(tenant_id) or {}).get("currency") or DEFAULT_CODE
    if code not in SUPPORTED:
        # A code that fell out of SUPPORTED must not leave the app with no
        # symbol; fall back rather than render bare numbers.
        log.warning("[currency] tenant %s has unsupported code %s", tenant_id, code)
        code = DEFAULT_CODE
    return {"code": code, **SUPPORTED[code]}


@router.get("")
def get_currency(user: CurrentUser = Depends(get_current_user)):
    """Readable by every role: it is needed to render any figure on any screen."""
    return ok({"current": currency_of(user.tenant_id),
               "supported": [{"code": c, **v} for c, v in SUPPORTED.items()]})


@router.patch("")
def set_currency(
    body: CurrencyUpdate,
    # Admin only. This relabels every figure the company has already loaded.
    user: CurrentUser = Depends(require_role("admin")),
):
    code = body.code.upper().strip()
    if code not in SUPPORTED:
        raise AppError(
            "currency_not_supported",
            f"Currency '{code}' is not supported.",
            status_code=400,
            params={"code": code, "supported": sorted(SUPPORTED)},
        )
    update_settings(user.tenant_id, {"currency": code})
    log.info("[currency] tenant %s -> %s (by %s)", user.tenant_id, code, user.user_id)
    return ok({"current": currency_of(user.tenant_id)})
